Wrapping compiler builtins

stoklund opened the pull request:
As [recently proposed on Zephyr-devel](https://lists.zephyrproject.org/g/devel/topic/31422932), this PR implements wrapper functions for a number of commonly used compiler builtins.

- CMake is used to detect which builtins are supported by the target compiler. A `<ccfeatures.h>` header file is generated.
- The new functions are provided in `<misc/math_extras.h>` with inline implementations that are either portable C or using compiler builtins when available.

I also replaced the use of naked overflow arithmetic builtins throughout the code. I have not yet done the same for CLZ/CTZ.

There are more candidates for wrapping like this:

- The `find_lsb_set()` and `find_msb_set()` functions could be implemented in terms of `__builtin_ffs()` and the existing bit-counting functions. It may also make sense to provide these in `size_t` variants, depending on how they are used.
- Byte swapping operations as currently provided by `<misc/byteorder.h>`.

Commit:
misc: Portable math_extras.h implementations.

Add a <misc/math_extras.h> header file with portable implementations of
a number of commonly used math and bit counting operations that are not
available in the C programming language.

The math_extras functions fall into two groups:

1. Unsigned integer arithmetic with overflow detection. Addition and
   multiplication are provides for the u32_t, u64_t, and size_t types.
   This commit doesn't include subtraction and division, and there are
   no signed operations. These could be added later if there is a need.

2. Bit-counting operations like clz, ctz, and ffs. These are provided
   for u32_t and u64_t only. I don't see a need for size_t operations,
   but they could be added if needed.

A follow-on commit will add more efficient implementations using
compiler builtins for those compilers that support it.

Another commit will replace other uses of "naked" compiler builtins with
calls to these functions.

Signed-off-by: Jakob Olesen <jolesen@fb.com>

Commit:
misc: Replace uses of __builtin_*_overflow() with <misc/math_extras.h>.

Use the new math_extras functions instead of calling builtins directly.

Change a few local variables to size_t after checking that all uses of
the variable actually expects a size_t.

Signed-off-by: Jakob Olesen <jolesen@fb.com>

andyross reviewed (changes requested):
Not really loving the way this inserts cmake into the lowest level of the build.  Build systems come and go like leaves on the wind.   And testing these all with per-app configure-style build-a-test-c-program seems terribly wasteful for a project that has IIRC 1700 such apps built for every CI pass.  And... what does this give us that we don't already have?  We know very well already which compilers support these builtins statically, there's zero need to probe this at build time.

FWIW: the abstracted API and generic fallback implementations look great to me.  It's the selection via complicated build magic that seems fragile and needless.

tyrantlizard77 reviewed (commented):
Overall LGTM, at least the interface definition and C implementations.
Is there a way to have two cases in math_extras testcase.yaml, one of which forces the C implementations to be exercised?

Commit:
misc: Always test the portable version of the math_extras.h functions.

The implementation of the <misc/math_extras.h> functions will use the
portable C code when PORTABLE_MISC_MATH_EXTRAS is defined before
including the header.

Use this feature macro to make sure that the unit test always tests the
portable functions in addition to the ones using builtins for the
current compiler.

Signed-off-by: Jakob Olesen <jolesen@fb.com>

Commit:
misc: Implement math_extras.h function with GCC builtins.

When available, use the GCC / Clang builtins to implement the
math_extras functions. Otherwise, use the portable versions.

Signed-off-by: Jakob Olesen <jolesen@fb.com>

stoklund commented:
> We know very well already which compilers support these builtins statically, there's zero need to probe this at build time.

I'm probably missing something obvious, but could you explain how that would work in practice?

I don't think compilers are publishing tables of which release first supported a given builtin, so you have to go through old release notes to figure that out. That is work, but doable of course.

When you test for compiler versions using macros like `__GNUC__`, so have to look out for compilers like Clang and ICC that masquerade as a certain GCC version by also defining this macro. There are cases of [ICC claiming to be compatible with a certain GCC version, but with some builtins missing](https://software.intel.com/en-us/forums/intel-c-compiler/topic/758013).

Clang has a `__has_builtin()` macro which does exactly what we need, but unfortunately only Clang has that, AFAIK.

I am also a little concerned about various toolchains from vendors. These are often based on GCC or Clang, but with a number of patches applied. This means that builtins can come and go while the compiler still claims to be some major version of GCC/Clang.

> Is there a way to have two cases in math_extras testcase.yaml, one of which forces the C implementations to be exercised?

I added a commit which does this.

jukkar commented on include/misc/math_extras_impl.h:
```diff
@@ -0,0 +1,97 @@
+{
+	int b = 0;
+
+	while (b < 32 && (x & 1) == 0)
```
Missing {} here

jukkar commented on include/misc/math_extras_impl.h:
```diff
@@ -0,0 +1,97 @@
+
+	while (b < 32 && (x & 1) == 0)
+		b++, x >>= 1;
+	return b;
```
empty line needed before the return statement

stoklund replied:
Thanks, I'll fix these lints.

pfalcon commented:
> The new functions are provided in <misc/math_extras.h>

I'd say, I find such a name unexpected. math.h is a well-known header which provides functions like sin() and cos(). Now I'd find strange to find functions like CLZ in math_extras.h. That said, I don't really have suggestion for better name (I'd apparently emphasize that it's wrappers for compiler builtin functions, with a name like cc_wrappers.h).

>  And testing these all with per-app configure-style build-a-test-c-program seems terribly wasteful

+1 for the concern of calling gazillion of check_c_source_compiles() which apparently spawn a subprocess each time.

SebastianBoe commented on Kconfig.zephyr:
```diff
@@ -250,6 +250,14 @@ config COMPILER_OPT
+	bool "Avoid using compiler builtin functions"
+	help
+	  When this option is set, Zephyr will avoid using the compiler's
+	  builtin functions for extra math operations. Portable C
```
> for extra math operations.

The option name implies something else. It implies we would avoid built-in memcmp etc. as well.

Perhaps the option name should be extended to disambiguate?

stoklund replied:
Yeah, I agree.

I'm not even sure if this should be a Kconfig option or just a feature macro (with a long name) that can be defined before including the header. It's really only useful for the unit test which wants to test the portable implementations.

SebastianBoe commented on cmake/ccfeatures.cmake:
```diff
@@ -0,0 +1,49 @@
+  }
+" HAVE___BUILTIN_CTZ)
+
+check_c_source_compiles("
```
This must be integrated with the ZEPHYR_TOOLCHAIN_CAPABILITY_CACHE to reach an acceptable cmake configure performance.

SebastianBoe commented:
> Not really loving the way this inserts cmake into the lowest level of the build.

I might misunderstand you, but be aware that we are already checking for compiler flag compatibility at build time. This is behaviour inherited from the Linux kernel (cc-option), and presumably exists to easily build the kernel with a wide range of gcc versions.

The use of #cmakedefine and configure_file is also in use, e.g. with version.h, and has proven to work well. It should be noted, that migrating to a different build system would only need porting of the *.h.in file, not all users of the .h file.

olofj commented:
@andyross

Linux does this at build time by detecting the compiler being used instead of having the build system do it. I.e. there's a <linux/compiler.h> that detects compiler based on defines. Nothing quite as elaborate is needed here yet, but the basic approach seems useful:

https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/include/linux/compiler_types.h#n58

I.e. #ifdef + per-compiler sub-include, and then in that include key off of version if needed.

Would that be an approach you'd support?

stoklund commented:
> > The new functions are provided in <misc/math_extras.h>
> 
> I'd say, I find such a name unexpected. math.h is a well-known header which provides functions like sin() and cos(). Now I'd find strange to find functions like CLZ in math_extras.h. That said, I don't really have suggestion for better name (I'd apparently emphasize that it's wrappers for compiler builtin functions, with a name like cc_wrappers.h).

Heh. I happen to be a mathematician, and I find the C standard committee's idea of "math" rather limiting ;-)

But seriously, the name comes from the LLVM and Mozilla projects which have `MathExtras.h` and `MathAlgorithms.h` headers respectively. The idea behind these names is that the headers provide little functions that just operate on numbers. The functions don't have any context other than numbers, hence "math".

An alternative is `<misc/int_extras.h>` since these are all missing integer operations. (Rust provides all of these ops [as part of their integer types](https://doc.rust-lang.org/std/primitive.u32.html)).

IMHO, the fact that many of these functions are implemented in terms of compiler builtins should just be an implementation detail. Somebody using this header is not going to care if they are calling portable C functions or builtins.

andyross commented:
FWIW, the cmake criticism was mostly aesthetics.  I'm not really interested in litgating that per se, the point is that I'd prefer the lowest level of our headers be free of tool interactions.  C code should be C code.

The focused complaint was the runtime testing, which is just a bad idea given the breadth and structure of our test suite.

The broader complaint is that we already have a framework for specifying toolchain configuration statically via ZEPHYR_TOOLCHAIN_VARIANT and the toolchain.h header tree, and we should be leveraging that if at all possible.  I just don't see the need to try to detect this at runtime, do we even **HAVE** toolchains that support these functions piecewise?  Seems like all we really care about is "gcc" and "other", and can just handle things the way we already do, no?

Commit:
toolchain: Define a HAS_BUILTIN(x) macro in toolchain.h.

Compilers based on Clang provide a __has_builtin(x) macro which can be
used to detect in the preprocessor if a given builtin function is
supported by the compiler.

For other compilers (notably GCC), we provide an alternative definition
of HAS_BUILTIN(x) that depends on the toolchain-specific header file to
declare which builtin functions are supported based on the current
compiler version.

Signed-off-by: Jakob Olesen <jolesen@fb.com>

stoklund commented:
Thanks for the reviews. I've pushed an updated version which:

* Doesn't use CMake to detect the presence of compiler builtins at configure time. Instead, it uses a similar method to the Linux kernel to detect builtins at compile time. The existing `<toolchain.h>` header now provides a `HAS_BUILTIN(x)` macro for the purpose.
* Doesn't add anything to `Kconfig.zephyr`. The new header file just detects the presence of a `PORTABLE_MISC_MATH_EXTRAS` feature macro to disable the use of compiler builtins for testing purposes.
* Fixes nits found by reviewers and tools.

mped-oticon commented on include/misc/math_extras_impl.h:
```diff
@@ -0,0 +1,187 @@
+
+	*result = c;
+
+	return a > 0 && c / a != b;
```
nit: Parens would be nice to make precedence clear.

andyross replied:
Alternative nit: "n != 0" better captures the logic.  You're avoiding the divide by zero edge case here.  My first thought on seeing that > was that you must have a overflow/signedness bug since a is unsigned.

mped-oticon reviewed (approved):
LGTM. This a welcome change that adds toolchain flexibility and rigour.

andyross commented on include/misc/math_extras_impl.h:
```diff
@@ -0,0 +1,187 @@
+	int b = 0;
+
+	while (b < 32 && (x & 1) == 0) {
+		b++, x >>= 1;
```
So... frankly I don't personally care, but that use of the comma operator is needless and I'm all but certain someone's coding style document (@ceolin can tell you all about the world he's been stuck in) disallows this.   Actually given the structure, maybe this should just be a `for()`  (where, heh, you could use the comma in a well-justified way because it has to fit in the final expression).

andyross commented on include/misc/math_extras_impl.h:
```diff
@@ -0,0 +1,187 @@
+#if use_builtin(__builtin_ctzll)
+static inline int u64_count_trailing_zeros(u64_t x)
+{
+	return x == 0 ? 64 : __builtin_ctzll(x);
```
Genuine confusion here.  Why are we making the semantics (for an argument of zero) of our own "count_trailing_zeros" function...  **DIFFERENT** than the "ctz" builtin that we're supposed to be wrapping?  I mean... the whole point to this API is to wrap the builtin!  It should do the same thing, right?

stoklund replied:
Here's how I look at it: The purpose of *this patch* is indeed to wrap compiler builtins and remove them from the general source code. However, looking beyond this PR, the purpose of the API in `math_extras.h` is to provide a set of commonly useful little math routines. For users of the API, it is not interesting whether they are implemented in portable C or with compiler builtins.

The x86 BSF/BSR instructions produce an undefined result for x=0, and GCC turned that into full-blown nasal demons C undefined behavior in their documentation of the ctz/clz builtins. Other instruction set architectures also have clz/ctz instructions, and they all agree that x=0 should return the size of the register. Intel even added LZCNT and TZCNT instructions that do the same.

So since there is a meaningful return value for x=0, I thought it would be better to return that instead of exposing users to nasal demons in a pretty surprising way.

Clang understands what's going on. Here's its armv7 code:

```asm
_u32_count_leading_zeros:
	clz	r0, r0
	bx	lr

_u32_count_trailing_zeros:
	rbit	r0, r0
	clz	r0, r0
	bx	lr
```

andyross replied:
Seems like this needs an explanatory comment then about the inconsistent edge case, and maybe a `#ifdef __i386` or whatnot for clarity.   While it's nice that clang optimizes this correctly on ARM, I can all but guarantee you that many of our toolchains won't.

stoklund replied:
Unfortunately, `__builtin_ctz(0)` is undefined behavior on *all* architectures, thanks to the way it is defined and documented.

This means that if your code contains a line with `__builtin_ctz(x)` anywhere, the compiler is permitted to assume that x != 0 anywhere it can prove that this line has been or will be executed:

```c
  int first_missile = __builtin_ctz(x);

  if (x == 0) {
      disarm_missiles();
  }

  launch_missiles();
```

Here, the compiler is permitted to remove the conditional. It doesn't matter which instruction it ends up using for the builtin.

Many people find this infuriating, and understandably so.

andyross replied:
Then... maybe __builtin_ctz() is just problematic and a better solution would be arch-level inline assembly wrappers?  I mean... if the tool is broken why are we using it?

olofj replied:
@andyross either way, isn't this patch a step in the right direction, in that it gives a single helper entrypoint such that if you want to replace it with an assembly stub later for a specific arch, you can? Meanwhile, using this instead of open-coding __builtin_foo() in the code is a clear incremental improvement.

andyross replied:
Fair enough.  I'll remove the -1.  Can we at least get a big comment on the usage here explaining that extra test and the fact that it's (hopefully) optimized away for the benefit of people like me would might be tempted to come in and "fix" it later?

andyross commented on include/toolchain.h:
```diff
@@ -15,6 +15,24 @@
+ */
+#ifdef __has_builtin
+#define HAS_BUILTIN(x) __has_builtin(x)
+#else
```
Why bother here?  We're defining the HAS_BUILTIN_xxx symbols unconditionally anyway, right?  So what's the value of conditionally falling back to a toolchain macro when we could just do the same thing the same way for everyone and eliminate these lines?

stoklund replied:
This case covers Clang-derived toolchains and [possibly future GCC versions](https://gcc.gnu.org/bugzilla/show_bug.cgi?id=66970).

I'm trying to cover a more open-ended set of toolchains, a bit like Oticon's efforts.

andyross commented on kernel/include/syscall_handler.h:
```diff
@@ -353,9 +354,9 @@ extern int z_user_string_copy(char *dst, const char *src, size_t maxlen);
-							      &product), \
+		Z_SYSCALL_VERIFY_MSG(!u32_mul_overflow((u32_t)(nmemb), \
+						       (u32_t)(size), \
+						       &product), \
```
Same deal.  Please segregate whitespace cleanup in separate commits.

andyross commented on kernel/poll.c:
```diff
@@ -273,11 +273,10 @@ Z_SYSCALL_HANDLER(k_poll, events, num_events, timeout)
+	if (Z_SYSCALL_VERIFY_MSG(!u32_mul_overflow(num_events,
+						   sizeof(struct k_poll_event),
+						   &bounds),
+				 "num_events too large")) {
```
This change just looks like whitespace muckery unrelated to the patch.  Probably best to remove chaff from patches that actually do nontrivial stuff.

andyross commented on kernel/thread.c:
```diff
@@ -475,9 +476,8 @@ Z_SYSCALL_HANDLER(k_thread_create,
-							     stack_size,
-							     &total_size),
+	Z_OOPS(Z_SYSCALL_VERIFY_MSG(!u32_add_overflow(K_THREAD_STACK_RESERVED,
+						      stack_size, &total_size),
```
More whitespace cleanup; should be in separate patch.

andyross commented on include/misc/math_extras_impl.h:
```diff
@@ -0,0 +1,108 @@
+	int b = 0;
+
+	while (b < 32 && (x >> 31) == 0) {
+		b++, x <<= 1;
```
Same note about comma operator.

andyross reviewed (changes requested):
I'm much more comfortable with the build integration, thanks.  A few style nitpicks that you can consider or ignore.

Just one question about the CTZ implementation is all I can find to whine about.

jukkar commented on include/misc/math_extras_impl.h:
```diff
@@ -0,0 +1,187 @@
+
+	*result = c;
+
+	return a > 0 && c / a != b;
```
It would be really helpfull if you could add () here and other returns above. Now it needs mental effort to read the line in order to figure out what it is doing.

olofj commented:
> The broader complaint is that we already have a framework for specifying toolchain configuration statically via ZEPHYR_TOOLCHAIN_VARIANT and the toolchain.h header tree, and we should be leveraging that if at all possible. I just don't see the need to try to detect this at runtime, do we even HAVE toolchains that support these functions piecewise? Seems like all we really care about is "gcc" and "other", and can just handle things the way we already do, no?

Not really, external toolchain can be of different versions, and different versions of GCC might or might not have these built-ins (depending on how recent it is). That was one of Jakob's original observations.

stoklund commented:
Updated patch:

- Clarify the precedence of arithmetic and relations with parentheses.
- Avoid the use of the comma operator and use a for loop instead of a while loop.
- Verified that all the patches are free of whitespace cleanup hunks.

mbolivar commented:
> FWIW, the cmake criticism was mostly aesthetics. I'm not really interested in litgating that per se, the point is that I'd prefer the lowest level of our headers be free of tool interactions. C code should be C code.

I find this comment frankly hilarious given the number of `CONFIG_` and `DT_` uses in our code.

stoklund commented:
Thanks @andyross. I've pushed a rebased version which also adds a big comment about the clz/ctz zero checks.
